var/spack/repos/builtin/packages/py-bbp-workflow/package.py

```diff
@@ -36,6 +36,8 @@ class PyBbpWorkflow(PythonPackage):
     depends_on('py-simwriter',         type='run')
     depends_on('py-voxcell',           type='run')
 
+    depends_on('py-h5py~mpi', type='run')
```

Comment:
This is brittle in that
* Spack uses dictionaries to store the dependency tree
* Spack uses RedHat's Python 3.4 with no guaranteed dictionary traversing order
* If another dependency requests the implicit `py-h5py+mpi` (since `+mpi` is the default), and is encountered first, you're screwed

IMO, it would be simpler to use `- py-bbp-workflow ^py-h5py~mpi` in `deploy/environments/applications.yaml`.

In the next iteration of our deployment, I'd like to address this a wee bit differently by providing different defaults for MPI variants for different teams.

Reply:
Thanks for the suggestion! I was not sure which would be the preferred way.

Reply:
Hmm, if I do `spack spec py-visual-cortex-analysis` I see `py-mpi4py` which was not there when using `depends_on('py-h5py~mpi', type='run')`. Any ideas?

Reply:
Don't see it?
```
❯ spack spec py-visual-cortex-analysis\^py-h5py~mpi
Input spec
--------------------------------
py-visual-cortex-analysis
    ^py-h5py~mpi

Concretized
--------------------------------
py-visual-cortex-analysis@develop%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-bmtk@0.0.8%gcc@9.3.0 arch=linux-rhel7-x86_64
        ^py-h5py@3.1.0%gcc@9.3.0~mpi arch=linux-rhel7-x86_64
            ^hdf5@1.10.7%gcc@9.3.0+cxx~debug~fortran+hl~java+mpi+pic+shared~szip~threadsafe api=none arch=linux-rhel7-x86_64
                ^hpe-mpi@2.22.hmpt%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-rhel7-x86_64
            ^py-cached-property@1.5.2%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^py-setuptools@50.3.2%gcc@9.3.0 arch=linux-rhel7-x86_64
                    ^python@3.8.3%gcc@9.3.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-rhel7-x86_64
            ^py-cython@0.29.21%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-numpy@1.19.4%gcc@9.3.0+blas+lapack arch=linux-rhel7-x86_64
                ^intel-mkl@2018.3.222%gcc@9.3.0~ilp64+shared threads=none arch=linux-rhel7-x86_64
            ^py-pkgconfig@1.5.1%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^pkgconf@1.7.3%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^py-nose@1.3.7%gcc@9.3.0 arch=linux-rhel7-x86_64
        ^py-jsonschema@3.2.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-attrs@20.3.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-pyrsistent@0.16.0%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^py-six@1.16.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-setuptools-scm@4.1.2%gcc@9.3.0+toml arch=linux-rhel7-x86_64
                ^py-toml@0.10.0%gcc@9.3.0 arch=linux-rhel7-x86_64
        ^py-matplotlib@3.3.3%gcc@9.3.0~animation~fonts+image~latex~movies backend=agg arch=linux-rhel7-x86_64
            ^freetype@2.10.1%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^libpng@1.6.37%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-certifi@2020.6.20%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-cycler@0.10.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-kiwisolver@1.1.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-pillow@7.2.0%gcc@9.3.0~freetype~imagequant+jpeg~jpeg2000~lcms~tiff~webp~webpmux~xcb+zlib arch=linux-rhel7-x86_64
                ^libjpeg-turbo@2.0.4%gcc@9.3.0 arch=linux-rhel7-x86_64
                    ^cmake@3.15.7%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt patches=1c540040c7e203dd8e27aa20345ecb07fe06570d56410a24a266ae570b1c4c39,bf695e3febb222da2ed94b3beea600650e4318975da90e4a71d6f31a6d5d8c3d arch=linux-rhel7-x86_64
                    ^nasm@2.15.05%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-pyparsing@2.4.2%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-python-dateutil@2.8.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^qhull@2015.2%gcc@9.3.0~ipo build_type=RelWithDebInfo patches=10ddc62a1600b0ede23fb0502e29dca389d18689fd8dfe7a849e3a989c0e607e arch=linux-rhel7-x86_64
        ^py-pandas@1.2.5%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-bottleneck@1.2.1%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-numexpr@2.7.0%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-pytz@2020.1%gcc@9.3.0 arch=linux-rhel7-x86_64
        ^py-scikit-image@0.17.2%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-imageio@2.5.0%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^ffmpeg@4.2.2%gcc@9.3.0~X~avresample+bzlib~drawtext+gpl~libaom~libmp3lame~libopenjpeg~libopus~libsnappy~libspeex~libssh~libvorbis~libwebp~libzmq~lzma~nonfree~openssl~sdl2+shared+version3 arch=linux-rhel7-x86_64
            ^py-networkx@2.4%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^py-decorator@4.4.2%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-pywavelets@1.1.1%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-scipy@1.5.4%gcc@9.3.0 arch=linux-rhel7-x86_64
                ^py-pybind11@2.5.0%gcc@9.3.0~ipo build_type=RelWithDebInfo arch=linux-rhel7-x86_64
            ^py-tifffile@2020.10.1%gcc@9.3.0 arch=linux-rhel7-x86_64
        ^py-sympy@1.4%gcc@9.3.0 arch=linux-rhel7-x86_64
            ^py-mpmath@1.1.0%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-frozendict@1.2%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-lazy@1.2%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-pyyaml@5.3.1%gcc@9.3.0+libyaml arch=linux-rhel7-x86_64
        ^libyaml@0.2.4%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-seaborn@0.11.1%gcc@9.3.0 arch=linux-rhel7-x86_64
    ^py-tqdm@4.59.0%gcc@9.3.0~notebook~telegram arch=linux-rhel7-x86_64
```

Also, seems like you'd rather want `^py-h5py~mpi ^hdf5~mpi`… or maybe change to `srun --mpi=none` in launch scripts?

Reply:
My final goal is not to have `mpi4py` otherwise `bmtk` starts doing MPI magic...

Reply:
Looks like the modules are fine.